packages/teams-js/src/internal/utils.ts

```diff
@@ -54,8 +70,9 @@ export function validateOrigin(messageOrigin: URL): boolean {
     return false;
   }
   const messageOriginHost = messageOrigin.host;
+  const validDomains = GlobalVars.validOrigins.length !== 0 ? GlobalVars.validOrigins : validOriginsFallback;
```

Comment:
Are there scenarios where apps will be using the fallback list for a few seconds before the 'getDomainsFromCDN' call completes? If so, is that okay? Would it be better to await the getDomainsFromCDN call so that we know for sure that the correct domains have been retrieved?

Reply:
I initially had it like that, but found that execution was able to be completed faster by handling it like this.  An inspection into why was while I was debugging, the fetch call would complete and finish translating the json into a validDomains object, but the Promise for `getValidDomainsFromCDN` would take longer to execute and notify the calling function that the domains have been acquired by a range of sometimes 50ms to 150ms longer (This is just local code execution). Translating this to using the GlobalVariable allowed us to save that execution time. This also enabled us to move the `getValidDomainsFromCDN` to earlier in our pipeline as well.

My assumption was that on really slow WiFi we would run into the issue you described, but even there where acquiring the CDN endpoint takes about "1.2 seconds" ("Simulated Slow 3G") it was still acquiring and waiting for the `getDomainsFromCDN` call to finish first. 

I do like your suggestion and I believe it may result in fewer errors, so I will experiment with using `await` again.

Reply:
I have some feedback along these lines so I'm going to leave it in this discussion thread:

If possible, we should avoid adding a new global (makes it harder to test, leaks abstraction details to the world, potentially makes tree shaking and other optimizations more difficult in the future).

Ideally the fact that the origins are sometimes coming from a CDN and sometimes a fallback should be completely be abstract from the rest of the code. The current design seems like it mostly does this, but it's a little hard to tell because the exact componentization boundary lines are blurry since these functions live in the middle of the `utils.ts` class which has a lot of unrelated stuff in it. Is "utils" the component? That doesn't seem quite right.

I think _ideally_ the design would be something like a function that looks like this:

`getValidDomainsList(): URL[]`

...then internally, that function can asynchronously query the CDN or get it from a fallback or switch it out depending on network latency/connectivity/etc. It can also "internally" cache list it retrieves from the network. It can log when it retrieved it, when it switched from the fallback, etc. The point is that no one outside of that function has any knowledge of any of these details, all they know is "call this function to get the list."

With the current design:
- The `validateOrigin` function has to know details like there are two possible places to check for the list
- Any code in the entire app is allowed to go poking at the global list directly and know how to interpret it, what it means it if it's empty, etc.
etc.

My suggestion would be to consider how best encapsulate all the implementation details behind a single function or class or file or something to prevent inevitable misuse by future developers. Ideally in a completely separate file dedicated only to validOrigin functionality --> but I also recognize that moving code to other files can cause widespread refactoring pain so such potential changes don't have to be in the same change.

The one bit of implementation detail that I could justifiably see "leaking" outside of that component is the function you want to call early to get the CDN fetching process started, even if the list isn't needed at the time that's kicked off. You could imagine creating a `prefetchValidOrigins` function potentially or just make the early code call your single "get the list function" and then just ignore the result (since it was only calling it to start the network requests)

Hope this was helpful and not confusing. 😅

Reply:
@AE-MS I think I understood what you were trying to say and made modifications to the code to reflect it. Here's hoping I didn't get confused somewhere along the way.